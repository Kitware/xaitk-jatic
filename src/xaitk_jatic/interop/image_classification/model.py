"""
This module provides the `JATICImageClassifier` class, an adapter for integrating JATIC-based
image classifiers with the SMQTK `ClassifyImage` interface. It enables the use of JATIC
protocol-based classifiers in pipelines that require the SMQTK interface.

Classes:
    JATICImageClassifier: Adapts a JATIC image classification model for compatibility with
                          SMQTK's `ClassifyImage` interface.

Dependencies:
    - maite.protocols.image_classification: For the JATIC image classification protocol.
    - numpy: For numerical operations.
    - smqtk_classifier: For SMQTK classifier interfaces and classification utilities.
"""

from collections.abc import Hashable, Iterator, Sequence

import maite.protocols.image_classification as ic
import numpy as np
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T
from smqtk_classifier.interfaces.classify_image import IMAGE_ITER_T, ClassifyImage
from typing_extensions import override


class JATICImageClassifier(ClassifyImage):
    """
    Adapter for the JATIC image classification protocol, implementing the SMQTK `ClassifyImage` interface.

    This adapter allows a JATIC protocol-based classifier to be used in SMQTK pipelines by transforming
    classification outputs into the expected format.

    Attributes:
        _classifier (ic.Model): The JATIC protocol-based image classifier instance.
        _ids (dict[int, Hashable]): A dictionary mapping label IDs to human-readable names.
        _img_batch_size (int): The number of images to process in a single batch.

    Methods:
        get_labels: Retrieve the human-readable class labels.
        classify_images: Perform classification on a batch of input images.
        get_config: Raise a `NotImplementedError` for serialization (not yet supported).
    """

    def __init__(
        self,
        classifier: ic.Model,
        ids: Sequence[int],
        img_batch_size: int = 1,
    ) -> None:
        """
        Initialize the JATICImageClassifier with a JATIC protocol-based classifier.

        Args:
            classifier (ic.Model): The JATIC protocol-based image classification model.
            ids (dict[int, Hashable]): A dictionary mapping label IDs to human-readable names.
            img_batch_size (int, optional): The number of images to process in a single batch. Defaults to 1.
        """
        self._classifier = classifier
        self._ids = sorted(ids)
        self._img_batch_size = img_batch_size

    @override
    def get_labels(self) -> Sequence[Hashable]:
        return self._ids

    @override
    def classify_images(self, img_iter: IMAGE_ITER_T) -> Iterator[CLASSIFICATION_DICT_T]:  # noqa: C901
        all_out = list()
        batch = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Transform outputs of a single batch
        def _generate_outputs(batch: Sequence[np.ndarray]) -> list[dict[int, float]]:
            predictions = np.asarray(self._classifier(batch))

            return [dict(enumerate(pred)) for pred in predictions]

        # Batch model passes
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                all_out.extend(_generate_outputs(batch))
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return iter(all_out)

    @override
    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require further implementation to do so.",
        )
