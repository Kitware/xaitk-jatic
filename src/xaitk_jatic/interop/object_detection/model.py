from collections.abc import Hashable, Iterable, Sequence
from typing import Dict, Tuple

import maite.protocols.object_detection as od
import numpy as np
from smqtk_detection import DetectImageObjects
from smqtk_image_io import AxisAlignedBoundingBox


class JATICDetector(DetectImageObjects):
    """Adapter for JATIC object detection protocol.

    :param detector: The JATIC protocol-based detector.
    :param id_to_name: Mapping from label IDs to names.
    :param img_batch_size: Image batch size for inference.
    """

    def __init__(
        self,
        detector: od.Model,
        ids: Sequence[int],
        img_batch_size: int = 1,
    ):
        self._detector = detector
        self._ids = sorted(ids)
        self._img_batch_size = img_batch_size

    def detect_objects(
        self,
        img_iter: Iterable[np.ndarray],
    ) -> Iterable[Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]:
        all_out = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Combine JATIC detections for same bbox into one DetectImageObject detection
        def _xform_dets(
            bboxes: Iterable[AxisAlignedBoundingBox],
            labels: np.ndarray,
            probs: np.ndarray,
        ) -> Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]:
            dets_dict: Dict[AxisAlignedBoundingBox, Dict[Hashable, float]] = dict()
            for box, label, prob in zip(bboxes, labels, probs):
                if box not in dets_dict:
                    dets_dict[box] = {la: 0.0 for la in self._ids}
                dets_dict[box][label] = prob

            return [(box, prob_dict) for box, prob_dict in dets_dict.items()]

        # Convert bboxes into AxisAlignedBoundingBoxes
        # Protocols specify X0, Y0, X1, Y1 format in v0.5.0
        def _xform_bbox(box: np.ndarray) -> AxisAlignedBoundingBox:
            return AxisAlignedBoundingBox(box[0:2], box[2:4])

        # Convert protocol-based detector output into DetectImageObject format
        def _generate_outputs(batch: Sequence[np.ndarray]) -> None:
            predictions = self._detector(batch)

            for pred in predictions:
                boxes = [_xform_bbox(box) for box in np.asarray(pred.boxes)]
                labels = np.asarray(pred.labels)
                scores = np.asarray(pred.scores)

                all_out.append(_xform_dets(bboxes=boxes, labels=labels, probs=scores))

        # Batch model passes
        batch = list()
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                _generate_outputs(batch)
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return all_out

    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require further implementation to do so.",
        )
