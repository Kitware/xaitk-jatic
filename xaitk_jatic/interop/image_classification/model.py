from typing import Dict, Hashable, Iterator, Sequence
import numpy as np

from smqtk_classifier.interfaces.classify_image import ClassifyImage, IMAGE_ITER_T
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T

import maite.protocols.image_classification as ic


class JATICImageClassifier(ClassifyImage):
    """
    Adapter for JATIC classifier protocol.

    :param classifier: The JATIC protocol-based classifier.
    :param id2name: Mapping from label IDs to names.
    :param img_batch_size: Image batch size for inference.
    """

    def __init__(
        self,
        classifier: ic.Model,
        id2name: Dict[int, Hashable],
        img_batch_size: int = 1
    ):
        self._classifier = classifier
        self._id2name = dict(sorted(id2name.items()))
        self._img_batch_size = img_batch_size

    def get_labels(self) -> Sequence[Hashable]:
        return [self._id2name[id] for id in sorted(self._id2name.keys())]

    def classify_images(
      self,
      img_iter: IMAGE_ITER_T
    ) -> Iterator[CLASSIFICATION_DICT_T]:
        all_out = list()
        batch = list()

        # Convert from channels last to channels first
        # Channels first is specified in protocols v0.5.0
        def _to_channels_first(img: np.ndarray) -> np.ndarray:
            if img.ndim < 3:
                return img
            return np.moveaxis(img, -1, 0)

        # Transform outputs of a single batch
        def _generate_outputs(batch: Sequence[np.ndarray]) -> None:
            predictions = np.asarray(self._classifier(np.asarray(batch)))

            for pred in predictions:
                all_out.append({self._id2name[id]: score for id, score in enumerate(pred)})

        # Batch model passes
        for img in img_iter:
            batch.append(_to_channels_first(img))

            if len(batch) == self._img_batch_size:
                _generate_outputs(batch)
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return iter(all_out)

    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require"
            "further implementation to do so."
        )
