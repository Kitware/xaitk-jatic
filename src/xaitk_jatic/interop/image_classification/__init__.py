"""Implementations to use XAITK for image classifcation in a maite compliant way"""
