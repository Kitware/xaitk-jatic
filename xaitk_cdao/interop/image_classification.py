from typing import Callable, Hashable, Iterable, Iterator, Sequence, Union, Optional
import numpy as np

from smqtk_classifier.interfaces.classify_image import ClassifyImage
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T

from maite.protocols import (
    ImageClassifier, HasLogits, HasProbs, HasScores, SupportsArray
)

from xaitk_cdao.utils.data_typing import to_numpy_array

from scipy.special import softmax  # type:ignore


class JATICImageClassifier(ClassifyImage):
    """
    Adapter for JATIC classifier protocol.

    :param classifier: The JATIC protocol-based classifier.
    :param preprocessor: Callable that takes a batch of data and returns a batch of data
        for any preprocessing before model inference.
    :param img_batch_size: Image batch size for inference.
    """

    def __init__(
        self,
        classifier: ImageClassifier,
        preprocessor: Optional[Callable[[SupportsArray], SupportsArray]] = None,
        img_batch_size: int = 1
    ):
        self._classifier = classifier
        self._labels = classifier.get_labels()
        self._preprocessor = preprocessor
        self._img_batch_size = img_batch_size

    def get_labels(self) -> Sequence[Hashable]:
        return self._labels

    def classify_images(
      self,
      img_iter: Union[np.ndarray, Iterable[np.ndarray]]
    ) -> Iterator[CLASSIFICATION_DICT_T]:
        all_out = list()
        batch = list()

        # Transform outputs of a single batch
        def _generate_outputs(batch: SupportsArray) -> None:
            if self._preprocessor:
                batch = self._preprocessor(batch)

            classifier_output = self._classifier(batch)

            # Get probabilities and transform to required output format
            if isinstance(classifier_output, HasLogits):
                all_logits = to_numpy_array(classifier_output.logits)
                for logits in all_logits:
                    all_out.append(dict(zip(self._labels, softmax(logits))))
            elif isinstance(classifier_output, HasProbs):
                all_probs = to_numpy_array(classifier_output.probs)
                for probs in all_probs:
                    all_out.append(dict(zip(self._labels, probs)))
            elif isinstance(classifier_output, HasScores):
                all_scores = to_numpy_array(classifier_output.scores)
                all_label_ids = to_numpy_array(classifier_output.labels)
                for scores, label_ids in zip(all_scores, all_label_ids):
                    out = {la: 0. for la in self._labels}
                    for score, label_id in zip(scores, label_ids):
                        out[self._labels[label_id]] = score
                    all_out.append(dict(out))
            else:
                raise ValueError("Unknown classifier output type")

        # Batch model passes
        for img in img_iter:
            batch.append(img)

            if len(batch) == self._img_batch_size:
                _generate_outputs(batch)
                batch = list()

        # Leftover batch
        if len(batch) > 0:
            _generate_outputs(batch)

        return iter(all_out)

    def get_config(self) -> dict:
        raise NotImplementedError(
            "Constructor arguments are not serializable as is and require"
            "further implementation to do so."
        )
