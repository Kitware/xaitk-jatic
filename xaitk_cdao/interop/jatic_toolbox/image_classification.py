from typing import Hashable, Iterable, Iterator, Sequence, Union
import numpy as np

from smqtk_classifier.interfaces.classify_image import ClassifyImage
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T

from jatic_toolbox.protocols import Classifier, HasLogits, HasProbs

from xaitk_cdao.utils.data_typing import to_numpy_array

from scipy.special import softmax  # type:ignore


class JATICImageClassifier(ClassifyImage):
    """
    Wrapper for JATIC classifier protocol.
    """

    def __init__(self, classifier: Classifier, labels: Sequence[Hashable]):
        self._classifier = classifier
        self._labels = labels

    def get_labels(self) -> Sequence[Hashable]:
        return self._labels

    def classify_images(
      self,
      img_iter: Union[np.ndarray, Iterable[np.ndarray]]
    ) -> Iterator[CLASSIFICATION_DICT_T]:
        for img in img_iter:
            classifier_output = self._classifier([img])

            if isinstance(classifier_output, HasLogits):
                probs = softmax(to_numpy_array(classifier_output.logits)[0])
            elif isinstance(classifier_output, HasProbs):
                probs = to_numpy_array(classifier_output.probs)[0]
            else:
                raise ValueError("Unknown classifier output type")

            yield dict(zip(self._labels, probs))

    def get_config(self) -> dict:
        return {
            "classifier": self._classifier,
            "labels": self._labels
        }
