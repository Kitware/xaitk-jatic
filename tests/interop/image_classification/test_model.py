from __future__ import annotations

from collections.abc import Hashable, Iterator, Sequence
from contextlib import AbstractContextManager
from unittest.mock import MagicMock

import maite.protocols.image_classification as ic
import numpy as np
import pytest
from smqtk_classifier.interfaces.classification_element import CLASSIFICATION_DICT_T
from smqtk_core.configuration import configuration_test_helper

from xaitk_jatic.interop.image_classification.model import JATICImageClassifier

rng = np.random.default_rng()


class TestJATICImageClassifier:
    dummy_id_to_name_1 = {0: "A", 1: "B", 2: "C"}
    expected_labels = [0, 1, 2]
    dummy_id_to_name_2 = {0: "D", 2: "D", 1: "E"}

    dummy_out = np.asarray([0, 0.15, 0.8])

    @pytest.mark.parametrize(
        ("classifier", "id_to_name", "img_batch_size", "expectation"),
        [
            (
                MagicMock(spec=ic.Model),
                dummy_id_to_name_1,
                3,
                pytest.raises(NotImplementedError, match=r"Constructor arg"),
            ),
        ],
    )
    def test_configuration(
        self,
        classifier: ic.Model,
        id_to_name: dict[int, Hashable],
        img_batch_size: int,
        expectation: AbstractContextManager,
    ) -> None:
        """Test configuration stability."""
        inst = JATICImageClassifier(classifier=classifier, ids=list(id_to_name.keys()), img_batch_size=img_batch_size)
        with expectation:
            configuration_test_helper(inst)

    @pytest.mark.parametrize(
        (
            "classifier_output",
            "id_to_name",
            "img_batch_size",
            "imgs",
            "expected_return",
        ),
        [
            (
                [dummy_out],
                dummy_id_to_name_1,
                1,
                [rng.integers(0, 255, (3, 256, 256), dtype=np.uint8)],
                [dict(zip(expected_labels, dummy_out))],
            ),
            (
                [dummy_out],
                dummy_id_to_name_1,
                1,
                [rng.integers(0, 255, (256, 256), dtype=np.uint8)],
                [dict(zip(expected_labels, dummy_out))],
            ),
            (
                [dummy_out] * 2,
                dummy_id_to_name_1,
                2,
                rng.integers(0, 255, (2, 256, 256), dtype=np.uint8),
                [dict(zip(expected_labels, dummy_out)), dict(zip(expected_labels, dummy_out))],
            ),
        ],
        ids=["single 3 channel", "single greyscale", "multiple images"],
    )
    def test_smoketest(
        self,
        classifier_output: ic.TargetBatchType,
        id_to_name: dict[int, Hashable],
        img_batch_size: int,
        imgs: np.ndarray | Sequence[np.ndarray],
        expected_return: Iterator[CLASSIFICATION_DICT_T],
    ) -> None:
        """Test that MAITE classifier output is transformed appropriately."""
        mock_classifier = MagicMock(spec=ic.Model, return_value=classifier_output)

        inst = JATICImageClassifier(
            classifier=mock_classifier,
            ids=list(id_to_name.keys()),
            img_batch_size=img_batch_size,
        )

        res = list(inst.classify_images(imgs))
        assert res == expected_return

    @pytest.mark.parametrize(
        ("id_to_name", "expected_labels"),
        [
            (dummy_id_to_name_1, expected_labels),
            (dummy_id_to_name_2, expected_labels),
        ],
    )
    def test_labels(self, id_to_name: dict[int, Hashable], expected_labels: Sequence[Hashable]) -> None:
        """Test that get_labels() returns the correct labels."""
        inst = JATICImageClassifier(classifier=MagicMock(spec=ic.Model), ids=list(id_to_name.keys()))

        assert inst.get_labels() == expected_labels
