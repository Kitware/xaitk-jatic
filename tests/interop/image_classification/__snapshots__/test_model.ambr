# serializer version: 1
# name: TestJATICImageClassifier.test_labels[id_to_name0-expected_labels0]
  list([
    0,
    1,
    2,
  ])
# ---
# name: TestJATICImageClassifier.test_labels[id_to_name1-expected_labels1]
  list([
    0,
    1,
    2,
  ])
# ---
# name: TestJATICImageClassifier.test_labels[id_to_name2-expected_labels2]
  list([
    1,
    2,
    4,
  ])
# ---
# name: TestJATICImageClassifier.test_labels[id_to_name3-expected_labels3]
  list([
    1,
    2,
    4,
  ])
# ---
# name: TestJATICImageClassifier.test_smoketest[multiple images]
  list([
    dict({
      0: 0.0,
      1: 0.15,
      2: 0.8,
    }),
    dict({
      0: 0.0,
      1: 0.15,
      2: 0.8,
    }),
  ])
# ---
# name: TestJATICImageClassifier.test_smoketest[non-consecutive labels]
  list([
    dict({
      1: 0.0,
      2: 0.15,
      4: 0.8,
    }),
  ])
# ---
# name: TestJATICImageClassifier.test_smoketest[single 3 channel]
  list([
    dict({
      0: 0.0,
      1: 0.15,
      2: 0.8,
    }),
  ])
# ---
# name: TestJATICImageClassifier.test_smoketest[single greyscale]
  list([
    dict({
      0: 0.0,
      1: 0.15,
      2: 0.8,
    }),
  ])
# ---
# name: TestJATICImageClassifier.test_smoketest[unsorted non-consecutive labels]
  list([
    dict({
      1: 0.0,
      2: 0.15,
      4: 0.8,
    }),
  ])
# ---
